File: tests/Simple.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where

import qualified Data.ByteString as B
import           System.Exit

import           ETerm
import           Process


main :: IO ()
main = do
  runAssert "test1" [] (EInteger 700)
  runAssert "test2" [] (EInteger 15)
  runAssert "test3" [] (EInteger 64)
  runAssert "inc" [EInteger 1] (EInteger 2)
  runAssert "inc" [EInteger 100] (EInteger 101)
  runAssert "inc" [EInteger 1000] (EInteger 1001)
  runAssert "inc" [EInteger 987654321] (EInteger 987654322)
  runAssert "adder5" (map EInteger [101..105]) (EInteger 515)
  runAssert "test_if" [EInteger 101] (EAtom (AtomNo 31))
  runAssert "test_if" [EInteger 5] (EAtom (AtomNo 30))
  runAssert "test_if" [EInteger 1] (EAtom (AtomNo 29))
  runAssert "test_if" [EInteger 3] (EAtom (AtomNo 28))
  runAssert "simple_fun" [] (EInteger 50)
  runAssert "test_fun" [] (EInteger 92)

runAssert :: B.ByteString -> [ETerm] -> ETerm -> IO ()
runAssert fun args expected = do
  putStrLn $ "Function " ++ show fun ++ " with args " ++ show args
  actual <- runBeam ["tests/simple.beam"] "simple" fun args
  if actual == Just expected
    then putStrLn "ok."
    else do
      putStrLn $ "In " ++ show fun ++ ", " ++ show args ++ ": Expected " ++ show expected ++ " but got " ++ show actual
      exitFailure

File: tests/simple.erl
-module(simple).

-export([test1/0, test2/0, test3/0, inc/1, adder5/5, test_string/0, test_if/1, simple_fun/0, fun_in_fun/1, test_fun/0]).

test1() -> 700.

adder5(A,B,C,D,E) -> A + B + C + D + E.

test2() -> adder5(1, 2, 3, 4 ,5).

test3() -> tre_mult(3).

inc(N) -> N + 1.

tre_mult(N) -> inc(N) * inc(N) * inc(N).

test_string() -> "hello_world_strings".

test_if(N) ->
  if N > 100 -> ok100;
     N > 4 -> ok4;
     N < 2 -> ok2;
     true -> ok_true
  end.

simple_fun() ->
  A = fun(Arg) -> Arg + 42 end,
  A(8).

fun_in_fun(A) ->
  B = fun(C) -> 5 + C() + A() end,
  3 + B(A).

test_fun() ->
  fun_in_fun(fun() -> 42 end).
